src/Fantomas/TokenParser.fs

```diff
@@ -547,11 +547,43 @@ let private (|InterpStringEndOrPartToken|_|) token =
 let escapedCharacterRegex =
     System.Text.RegularExpressions.Regex("(\\\\(a|b|f|n|r|t|u|v|x|'|\\\"|\\\\))+")
 
-let rec private (|EndOfInterpolatedString|_|) tokens =
+let private splitAround<'a> (n: int) (xs: 'a list) : ('a list * ('a * 'a list) option) option =
+    let rec go (n: int) (heads: 'a list) (rest: 'a list) =
+        if n = 0 then
+            match rest with
+            | [] -> Some(List.rev heads, None)
+            | x :: rest -> Some(List.rev heads, Some(x, rest))
+        else
+            match rest with
+            | [] -> None
+            | x :: rest -> go (n - 1) (x :: heads) rest
+
+    go n [] xs
+
+let private (|EndOfInterpolatedString|_|) tokens =
     match tokens with
-    | StringTextToken stToken :: InterpStringEndOrPartToken endToken :: rest -> Some([ stToken ], endToken, rest)
-    | StringTextToken stToken :: EndOfInterpolatedString (stringTokens, endToken, rest) ->
-        Some(stToken :: stringTokens, endToken, rest)
+    | StringTextToken _ :: rest ->
+        let maybeLastTokenIndex =
+            rest
+            |> Seq.takeWhile
+                (function
+                | StringTextToken _
+                | InterpStringEndOrPartToken _ -> true
+                | _ -> false)
+            |> Seq.tryFindIndex
+                (function
+                | InterpStringEndOrPartToken _ -> true
+                | _ -> false)
+
+        match maybeLastTokenIndex with
+        | None -> None
+        | Some lastTokenIndex ->
+            match tokens |> splitAround (lastTokenIndex + 1) with
+            | Some (tokens, Some (endToken, rest)) ->
+                match tokens with
+                | [] -> None
+                | _ -> Some(tokens, endToken, rest)
+            | _ -> None
```

Comment:
I wanted to simplify this as follows:

```fsharp
| Some (head :: tail, Some (endToken, rest)) ->
    Some (head :: tail, endToken, rest)
| _ -> None
```

However, the need to re-concat `head` and `tail` immediately after pattern matching them felt a bit awkward and I'm not sure of perf implications. Any thoughts?

Reply:
You can match as follows if you like:

```fsharp
| Some (head :: tail as list, Some (endToken, rest)) ->
    Some (list, endToken, rest)
```

Reply:
Oh, that's perfect - thanks!